Add Support for TUN-based Channels (kqueue/epoll)

HeikoBornholdt opened the pull request:
## Motivation

I want to develop a [mesh VPN](https://github.com/drasyl-overlay/drasyl). Since I love netty, I would love to use it for this project. Therefore I need a [TUN](https://en.wikipedia.org/wiki/TUN/TAP)-based `Channel` implementation.

## Modifications

* Added native methods to `io.netty.channel.kqueue.BsdSocket` and `io.netty.channel.epoll.LinuxSocket`, allowing the creation of TUN devices and setting/getting the network interface MTU on macOS and Linux.
* Implemented `io.netty.channel.kqueue.KQueueTunChannel` and `io.netty.channel.epoll.EpollTunChannel`, allowing the creation of TUN-based channels using epoll and kqueue.
* Defined and implemented `io.netty.channel.socket.TunChannelConfig`, allowing setting/getting the network interface MTU.
* Modified `io.netty.channel.unix.Socket` to avoid `Socket#isIPv6(fd)` call that is not supported on TUN sockets.
* Modified `io.netty.channel.epoll.AbstractEpollChannel` allowing to set `AbstractEpollChannel#local` in `EpollTunChannel#doBind` as `io.netty.channel.kqueue.Native#epollCtlAdd` must be called after TUN device has been bound.
* Renamed `io.netty.channel.kqueue.AbstractKQueueDatagramChannel` to `io.netty.channel.kqueue.AbstractKQueueMessageChannel` because `KQueueTunChannel` needs to implement this class but is not datagram related. As this rename may break the API, I re-added `AbstractKQueueDatagramChannel` for legacy reasons.
* Added `io.netty.channel.socket.Tun4Packet` and `io.netty.channel.socket.Tun6Packet` to help work with IPv4 and IPv6 packets.
* Added `io.netty.example.tun.TunPingDevice` example creating a TUN device that will reply to IPMC/IPv6-ICMP echo ping requests.
* Added `io.netty.example.tun.TunEchoDevice` example that echoes all received IP packets (e.g., suitable for performance tests).
* Added native dependencies to `example/pom.xml` (might require changes to your release workflow).
* Modified `run-example.sh` to activate platform-dependent native transport maven profile.

## Result

Support for TUN-based `Channel`s on kqueue- and epoll-enabled platforms.

## Additional Nodes
If desired, I can also provide a `TunChannel` for Windows based on [Wintun](https://www.wintun.net/). [I already prototyped it](https://github.com/drasyl-overlay/netty-tun), but I'm not sure if the Wintun license is compatible with netty.

hyperxpro commented:
This is something that I'd love to work on. I'll review in sometime and get back to you.

hyperxpro commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollTunChannel.java:
```diff
@@ -0,0 +1,332 @@
+            return UnixChannelUtil.isBufferCopyNeededForWrite(buf) ? newDirectBuffer(buf) : buf;
+        }
+
+        throw new UnsupportedOperationException(
```
```suggestion
        throw new UnsupportedMessageTypeException(
```

HeikoBornholdt replied:
The `io.netty.handler.codec.UnsupportedMessageTypeException` class is part of `netty-codec` which creates a new dependency.

hyperxpro replied:
Alright, let's be consistent with the old API.

hyperxpro reviewed (changes requested):
First round of small changes

KevinRoebert commented on example/src/main/java/io/netty/example/tun/Echo6Handler.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx,
+                                Tun6Packet packet) throws Exception {
+        // swap source and destination addresses. Depending on the layer 4 protocol used, this may
```
But IP is a layer 4 protocol

hyperxpro replied:
Oops. I assumed it IP version. My bad. ;/

KevinRoebert replied:
Did the same mistake the first time I read this class ^^"

hyperxpro commented:
@chrisvest Can you please trigger CI?

HeikoBornholdt commented on transport-classes-kqueue/src/main/java/io/netty/channel/kqueue/KQueueTunChannel.java:
```diff
@@ -0,0 +1,237 @@
+            return UnixChannelUtil.isBufferCopyNeededForWrite(buf) ? newDirectBuffer(buf) : buf;
+        }
+
+        throw new UnsupportedOperationException(
```
The `io.netty.handler.codec.UnsupportedMessageTypeException` class is part of `netty-codec` which creates a new dependency.

HeikoBornholdt replied:
Moreover, throwing `UnsupportedOperationException` is consistent with the behavior of other `AbstractChannel#filterOutboundMessage` implementations.

Therefore I would suggest to stick with `UnsupportedOperationException`.

HeikoBornholdt commented on transport-classes-kqueue/src/main/java/io/netty/channel/kqueue/BsdSocket.java:
```diff
@@ -252,4 +264,69 @@ private static native int connectx(
+                throw newIOException("bind", res);
+            }
+        } else {
+            throw new Error("Unexpected SocketAddress implementation " + socketAddress);
```
I wanted to be consistent with [Socket#bind](https://github.com/netty/netty/blob/4.1/transport-native-unix-common/src/main/java/io/netty/channel/unix/Socket.java#L387) and [Socket#connect](https://github.com/netty/netty/blob/4.1/transport-native-unix-common/src/main/java/io/netty/channel/unix/Socket.java#L348). But if you still like me to change it, I will do it :) .

hyperxpro replied:
That's fine then.

hyperxpro commented:
I'll do benchmarking and testing today.

hyperxpro commented on transport/src/main/java/io/netty/channel/socket/TunAddress.java:
```diff
@@ -0,0 +1,53 @@
+        this.ifName = ifName;
+    }
+
+    public TunAddress() {
```
Can we add JavaDoc here? So it's clear whether we can pass 'null' if we don't already have a Tun interface?

hyperxpro commented on transport/src/main/java/io/netty/channel/socket/TunAddress.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public String toString() {
+        if (ifName == null) {
+            return "";
+        } else {
+            return ifName;
+        }
```
return "Interface: " + ifName;
```suggestion
        return "Tun Interface: " + ifName;
```

hyperxpro commented on transport/src/main/java/io/netty/channel/socket/TunPacket.java:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @return the IP version.
+     */
+    public abstract int version();
```
How about using an `enum`?

HeikoBornholdt replied:
I can add this enum that I made for another project: https://github.com/drasyl-overlay/netty-tun/blob/master/src/main/java/org/drasyl/channel/tun/InetProtocol.java

hyperxpro replied:
That looks good.

HeikoBornholdt replied:
As this method returns the IP version, not a layer 4 protocol number. Therefore, I changed the method to return a [`InternetProtocolFamily`](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/socket/InternetProtocolFamily.java) enum.

hyperxpro commented on transport/src/main/java/io/netty/channel/socket/TunPacket.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Returns the destination address.
+     *
+     * @return the destination address.
```
We can remove this line

hyperxpro commented on transport/src/main/java/io/netty/channel/socket/TunPacket.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Returns the source address.
+     *
+     * @return the source address.
```
We can remove this line

hyperxpro reviewed (changes requested):
Can we add test cases that test actual traffic flow because the current one only test packet level? Can we do an ICMP ping or a UDP echo?

hyperxpro commented:
This works like charm. Successfully built a bridge-tunnel connection between 2 EC2.

But my main concern is regarding performance since we are only using 1 thread for handling all I/O. Is there any way?

hyperxpro commented:
See https://docs.kernel.org/networking/tuntap.html#multiqueue-tuntap-interface

HeikoBornholdt commented:
> Can we add test cases that test actual traffic flow because the current one only test packet level? Can we do an ICMP ping or a UDP echo?

I will try to implement an ICMP ping test

hyperxpro commented:
You can use the UDP `SO_REUSEPORT` approach. We set this socket option and bind on the same port multiple times but the Handler remains the same for all channels. Each time we bind, we use the next `EventLoop` on `EventLoopGroup`.

You can use multi-queue to accept multiple packet streams using different channels but on the same Handler.

HeikoBornholdt commented:
I followed your advice and added `EpollTunChannelOption.IFF_MULTI_QUEUE`. If set to `true`, one can create multiple `TunChannel`s assigned to the same TUN device. I also expanded the `TunEchoDevice` example so one can define the number of channels to be created.

Earlier, you suggested writing some tests: Have you thought of a fully automated test (e.g., JUnit) or a new testsuite Maven module to help with manually testing?

hyperxpro commented:
Looks good but we need more test cases to cover all functionality.

hyperxpro commented:
We don't need an extra module I guess.

HeikoBornholdt commented:
I am thinking on an integration test creating a `TunChannel` with `Ping4Handler` added to it.
In addition, we need a `DatagramChannel` sending an ICMP Ping Request and waiting for a response.
This must be made for both epoll and kqueue. The kqueue test would require us to create a new macOS-based GitHub Workflows.

Is this in line with your expectation?

hyperxpro commented:
@normanmaurer Is it fine if we create a workflow for testing this on macOS?

HeikoBornholdt commented:
* Added check if `IFF_MULTI_QUEUE` is available when building `netty-transport-native-epoll`.
* Added `io.netty.channel.epoll.Native.IS_SUPPORTING_MULTI_QUEUE`, allowing to check if `IFF_MULTI_QUEUE` feature is available.
* Added tests for epoll and kqueue (currently, only UDP-based traffic is processed. I don't see any advantages of adding a TCP-based test as well, as it makes no difference for the TUN device. But if you insist, I can add it).
* Adjusted `docker/docker-compose.yaml` to make TUN tests work in CI.

hyperxpro commented:
@chrisvest Please add this to your TO-DO :)

chrisvest commented:
This seems a bit exotic, so I'm wondering if we can get by with the minimum JNI changes in core Netty, and have the rest of the implementation in a contrib repository? It's also a big change in general, which makes me wonder if Netty 5 is a better target (fewer versions to maintain; would have to be ported anyway).

hyperxpro commented:
Sounds good.

@chrisvest If you can review the code then I can create a contrib repository and begin working with @HeikoBornholdt on porting this.

HeikoBornholdt commented:
I would love to see TUN support added to Netty 4.1.
Porting this feature to Netty 5 is already something I want to address next.

hyperxpro commented:
I'll be closely maintaining this module with @HeikoBornholdt if merged in 4.1. You can rely on us for maintenance.

KevinRoebert commented:
> I'll be closely maintaining this module with @HeikoBornholdt if merged in 4.1. You can rely on us for maintenance. @chrisvest

**+1** As a co-author, I am also willing to maintain the module.

hyperxpro commented:
@normanmaurer @chrisvest Can we do this folks?

AriseFX commented:
I would like to know the progress of this thing as I happen to have a need to develop a TUN interface :)

hyperxpro commented:
@chrisvest Gentle ping... :)

HeikoBornholdt commented:
Moin @normanmaurer 👋,

I remember your delight atsimilar PR https://github.com/netty/netty/pull/6346#issuecomment-279130473.
Any chance you could take a look at this PR?

KevinRoebert commented:
@normanmaurer @hyperxpro Is there any update on the review? Anything we can help with?

normanmaurer commented:
Unfortunately I dont feel we have the resources at the moment to maintain this and so put it in the main project. The best bet  for now would be to either make it part of the Netty-incubator or make it a separate repository under your own username.

KevinRoebert commented:
@normanmaurer @hyperxpro Do we get the okay for that?

hyperxpro commented:
We are having an internal discussion on this. Will get back soon with an update.

hyperxpro commented:
There you go folks: https://github.com/netty-contrib/transport-tun

Please do a PR and tag everyone for review.

hyperxpro commented:
Are you guys still interested in this?

KevinRoebert commented:
> Are you guys still interested in this?

Thank you for asking! We're definitely still interested in implementing this in the [netty-contrib/transport-tun](https://github.com/netty-contrib/transport-tun) repository. However, since some of the required Netty classes are `final` and package-private, we'd end up duplicating a significant amount of existing Netty code. This would mean we'd need to backport changes whenever the original code is updated. Implementing this feature in a separate repository with its own namespace would require a substantial amount of additional development and maintenance work and result in thousands of lines of duplicate code. Here are the specific duplicates we currently have:
* Netty C code (netty_epoll_linuxsocket.c, netty_epoll_native.c, netty_kqueue_bsdsocket.c) and all related Java classes that handle this C code.
* Creation of maven modules very similar to: netty-transport-native-epoll, netty-transport-native-kqueue, and netty-transport-unix-common.
* Netty CI setup for building the native code.
* The following Netty internal classes and their dependencies: AbstractEpollChannel.java, LinuxSocket.java, Native.java, AbstractKQueueChannel.java, AbstractKQueueDatagramChannel.java, BsdSocket. java, and unix/Socket.java.

We understand the reasoning behind outsourcing this feature to an additional module, but we believe direct integration into the Netty core repository would be preferable given the challenges mentioned.

chrisvest commented:
@hyperxpro What breakages does this need in 4.2?

bgmoon commented:
I'm very interested in this pull request as well.  We have been working with @HeikoBornholdt as well on the stand alone implementation.  However that implementation lacks the native transport integration that this PR would provide and that is necessary for any implementation on MacOS (OS/X) as apple has made some POSIX changes in latest versions that would require the native transport changes to overcome.  If there is anything we can do to support this effort and get it into 4.x, let us know.  We would very much like to use this code in a commercial product soonest.  BillyMoon CTO and Founder @ WhiteStar Communications, Inc.
